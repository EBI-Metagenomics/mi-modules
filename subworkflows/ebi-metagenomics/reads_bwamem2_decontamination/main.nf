include { BWAMEM2_MEM     } from '../../../modules/ebi-metagenomics/bwamem2/mem/main'
include { SAMTOOLS_BAM2FQ } from '../../../modules/ebi-metagenomics/samtools/bam2fq/main'


workflow READS_BWAMEM2_DECONTAMINATION {

    take:
    ch_reads       // channel (mandatory): [ val(meta), [ path(reads) ] ]
    ch_reference   // channel (mandatory): [ val(meta2), path(ref_index) ] | meta2 contains the name of the reference genome


    main:

    ch_versions = Channel.empty()

    BWAMEM2_MEM(ch_reads, ch_reference)
    ch_versions = ch_versions.mix(BWAMEM2_MEM.out.versions.first())

    BWAMEM2_MEM.out.bam
        .multiMap { meta, bam, bai ->
            bam: [ meta, bam, meta.single_end == false ]
        } 
        .set { ch_in_bam2fq }

    SAMTOOLS_BAM2FQ( ch_in_bam2fq.bam )
    ch_versions = ch_versions.mix(SAMTOOLS_BAM2FQ.out.versions.first())
   
    emit:
    decontaminated_reads = SAMTOOLS_BAM2FQ.out.reads  // channel: [ val(meta), [ path(decont_reads) ]]
    versions = ch_versions                            // channel: [ versions.yml ]

}



