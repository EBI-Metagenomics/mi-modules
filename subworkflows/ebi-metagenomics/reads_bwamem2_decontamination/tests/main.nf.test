nextflow_workflow {

    name "Test Subworkflow READS_BWAMEM2_DECONTAMINATON"
    script "../main.nf"
    workflow "READS_BWAMEM2_DECONTAMINATION"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/reads_bwamem2_decontamination"
    tag "bwamem2/mem"
    tag "samtools/bam2fq"

    test("paired_end decontamination") {
        when {
            workflow {
                """
                input[0] = Channel.of([
			[ id:'test' ], // meta map
			[ file('/hps/nobackup/rdf/metagenomics/service-team/projects/finding-pheno/shallow_shotgun/nf_tool/decont_subwf_test/local_data_test/input_reads/mock_1_R1.fastq.gz', checkIfExists: true),
			  file('/hps/nobackup/rdf/metagenomics/service-team/projects/finding-pheno/shallow_shotgun/nf_tool/decont_subwf_test/local_data_test/input_reads/mock_1_R2.fastq.gz', checkIfExists: true)] // list of raw-reads
                ])

                input[1] = Channel.fromPath("/hps/nobackup/rdf/metagenomics/service-team/projects/finding-pheno/shallow_shotgun/nf_tool/decont_subwf_test/local_data_test/database_indices/MGYG000317500.fna*", checkIfExists: true).collect().map { db_files -> [ [ ["id": "MGYG000317500.fna"] ], db_files ]}

                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out.decontaminated_reads[0][1].findAll {
                    file(it).name != "test_1.bwa.fq.gz" &&
                    file(it).name != "test_2.bwa.fq.gz"  }).match()},
                { assert file(workflow.out.decontaminated_reads[0][1].find {
                    file(it).name == "test_1.bwa.fq.gz" }).exists() },
                { assert file(workflow.out.decontaminated_reads[0][1].find {
                    file(it).name == "test_2.bwa.fq.gz" }).exists() },
                { assert snapshot(workflow.out.versions).match("versions") }
            )
        }
    }
}
