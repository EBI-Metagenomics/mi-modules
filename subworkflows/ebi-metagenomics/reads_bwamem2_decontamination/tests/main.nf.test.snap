{
    "versions": {
        "content": [
            [
                
            ]
        ],
        "timestamp": "2024-02-22T12:02:01.880623843"
    }
}